api/envoy/extensions/filters/http/api_key_auth/v3/api_key_auth.proto

```diff
@@ -28,27 +26,66 @@ option (xds.annotations.v3.file_status).work_in_progress = true;
 // .. code-block:: yaml
 //
 //    authentication_header: "X-API-KEY"
-//    keys:
-//      inline_string: |-
-//        clientID1:apiKey1
-//        clientID2:apiKey2
+//    credentials:
+//      entries:
+//      - api_key: 09876abcdefg
+//        client_id: user
+//
+// If multiple fields of ``authentication_header``, ``authentication_query``, and ``authentication_cookie``
+// are set. Then filter will try to fetch the API key from the following fields in order:
+//
+// 1. ``authentication_header``
+// 2. ``authentication_query``
+// 3. ``authentication_cookie``
 //
-message APIKeyAuth {
-  // keys used to authenticate the client.
-  // It should be a map of clientID to apiKey.
-  // The clientID serves solely for identification purposes and isn't used for authentication.
-  config.core.v3.DataSource keys = 1 [(udpa.annotations.sensitive) = true];
+// If none of the fields are set, then the filter will reject the request.
+message ApiKeyAuth {
+  message Credential {
+    // The value of the unique API key.
+    string api_key = 1 [(validate.rules).string = {min_len: 1}];
+
+    // The unique id or identity that used to identify the client or consumer.
+    string client_id = 2 [(validate.rules).string = {min_len: 1}];
+  }
+
+  message Credentials {
+    repeated Credential entries = 1;
+  }
+
+  // Api credentials used to authenticate the clients.
+  Credentials credentials = 1 [(udpa.annotations.sensitive) = true];
```

Comment:
The api key credenticals has clear structure. It would be better to use a typed message rather than DataSource.

It's OK to change the API because the #36709 was just merged.

Reply:
@adisuissa Yeah. I know. I prefer the explicitly defined structures rather than plain string or bytes. There are two possible compromised ways:

1. Use the DataSource as config field type but require it's content to keep the structure of the `Crendentials`. We can load it by the `loadFromJsonString` directly to a `Crendentials` proto type.
2. Add a new DataSource config field in the future if some require env/file support?

Both are ok to me, which one would you prefer?